### DISCUSSION TITLE
Row で select した情報の扱いについて

### DISCUSSION THREAD

**Discussion by ironclad-15256304** [domaframework/doma#998] [Q&A] [answered]
### 相談内容

いつも大変お世話になっております。

Criteria API を使用した SQL の構築時、取得したい情報の数が 10 以上である場合には `Row` による取得が有効だと認識しています。
`Row` による列の取得について調べてみたところ、自分の調べ方が悪いのか、 `Row` で情報を select することが可能である記述までは見つかるものの、情報を取得した後の扱いまで実装方法が明示されているドキュメントを見つけられませんでした。
（参考にしたサイト：[Doma入門 - Criteria API チートシート #Java - Qiita](https://qiita.com/nakamura-to/items/04def9dc3d218d50c71f), [Welcome to Doma 2 - Doma documentation](https://doma.readthedocs.io/en/latest/)）

`Row` に格納されている値は `LinkedHashMap` で実装されているようでしたので、単純に `values()` 等で値をループして index を元に情報を特定したりすればよいものと認識したのですが、こちらの認識で問題ないでしょうか。

**Comment by nakamura-to** [comment, marked as answer]
`Row`の`get`メソッドを使って値を取得できます。

```java
Employee_ e = new Employee_();

List<Row> list =
    nativeSql
        .from(e)
        .orderBy(c -> c.asc(e.employeeId))
        .select(e.employeeId, e.employeeName, e.employeeNo, ... 略 ...)
        .fetch();

for (Row row : list) {
    int id = row.get(e.employeeId);
    String name = row.get(e.employeeName);
}
```

**Reply by ironclad-15256304** [reply]
なるほど、たしかに事前に `EntityMetaModel` とその `PropertyMetaModel` が定義されているので、それをキーにそのまま `get` の条件に使用できるのですね。
取得したデータをそのまま返したい場合でも、メソッド内でその形式で取得したものを何かしらの型に詰め替えて返してあげればよいのですね。

ということは、 `Expressions` で記載された列については、

```
PropertyMetamodel<Integer> maxAge = Expressions.sum(e.age);
```

のような形で事前に定義したものを `select` , `get` 両方に指定してあげればよいということでしょうか。

**Reply by nakamura-to** [reply]
事前定義してもいいですし、同じ式（ここでは `sum(e.age)` ）を繰り返しても大丈夫です。

```java
List<Row> list = nativeSql
        .from(e)
        .orderBy(c -> c.asc(e.employeeId))
        .select(sum(e.age))
        .fetch();
...

for (Row row : list) {
    int age = row.get(sum(e.age));
}
```

**Reply by ironclad-15256304** [reply]
承知しました。迅速なご回答有難う御座います！！！！
